World Loading Issue on Startup

```java
18:39:37 [ERROR] nukkit.level.loadError
cn.nukkit.utils.LevelException: Constructor of class cn.nukkit.level.format.anvil.Anvil failed
        at cn.nukkit.level.Level.lambda$new$1(Level.java:314) ~[server.jar:?]
        at cn.nukkit.level.Level.<init>(Level.java:333) ~[server.jar:?]
        at cn.nukkit.level.Level.<init>(Level.java:302) ~[server.jar:?]
        at cn.nukkit.Server.loadLevel(Server.java:2136) [server.jar:?]
        at cn.nukkit.level.EnumLevel.initLevels(EnumLevel.java:23) [server.jar:?]
        at cn.nukkit.Server.<init>(Server.java:746) [server.jar:?]
        at cn.nukkit.Nukkit.main(Nukkit.java:120) [server.jar:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_272]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_272]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_272]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_272]
        at cn.nukkit.level.Level.lambda$new$1(Level.java:312) ~[server.jar:?]
        ... 6 more
Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
        at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:240) ~[?:1.8.0_272]
        at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158) ~[?:1.8.0_272]
        at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:117) ~[?:1.8.0_272]
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:246) ~[?:1.8.0_272]
        at java.io.BufferedInputStream.read(BufferedInputStream.java:265) ~[?:1.8.0_272]
        at java.io.DataInputStream.readByte(DataInputStream.java:265) ~[?:1.8.0_272]
        at cn.nukkit.nbt.stream.NBTInputStream.readByte(NBTInputStream.java:64) ~[server.jar:?]
        at cn.nukkit.nbt.tag.Tag.readNamedTag(Tag.java:91) ~[server.jar:?]
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:93) ~[server.jar:?]
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:89) ~[server.jar:?]
        at cn.nukkit.nbt.NBTIO.readCompressed(NBTIO.java:127) ~[server.jar:?]
        at cn.nukkit.level.format.generic.BaseLevelProvider.<init>(BaseLevelProvider.java:58) ~[server.jar:?]
        at cn.nukkit.level.format.anvil.Anvil.<init>(Anvil.java:41) ~[server.jar:?]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_272]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_272]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_272]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_272]
        at cn.nukkit.level.Level.lambda$new$1(Level.java:312) ~[server.jar:?]
        ... 6 more
18:39:37 [INFO ] No level called "nether" found, creating default nether level.
18:39:37 [ERROR] nukkit.level.loadError
cn.nukkit.utils.LevelException: Constructor of class cn.nukkit.level.format.anvil.Anvil failed
        at cn.nukkit.level.Level.lambda$new$1(Level.java:314) ~[server.jar:?]
        at cn.nukkit.level.Level.<init>(Level.java:333) ~[server.jar:?]
        at cn.nukkit.level.Level.<init>(Level.java:302) ~[server.jar:?]
        at cn.nukkit.Server.loadLevel(Server.java:2136) [server.jar:?]
        at cn.nukkit.level.EnumLevel.initLevels(EnumLevel.java:37) [server.jar:?]
        at cn.nukkit.Server.<init>(Server.java:746) [server.jar:?]
        at cn.nukkit.Nukkit.main(Nukkit.java:120) [server.jar:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_272]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_272]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_272]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_272]
        at cn.nukkit.level.Level.lambda$new$1(Level.java:312) ~[server.jar:?]
        ... 6 more
Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
        at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:240) ~[?:1.8.0_272]
        at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158) ~[?:1.8.0_272]
        at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:117) ~[?:1.8.0_272]
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:246) ~[?:1.8.0_272]
        at java.io.BufferedInputStream.read(BufferedInputStream.java:265) ~[?:1.8.0_272]
        at java.io.DataInputStream.readByte(DataInputStream.java:265) ~[?:1.8.0_272]
        at cn.nukkit.nbt.stream.NBTInputStream.readByte(NBTInputStream.java:64) ~[server.jar:?]
        at cn.nukkit.nbt.tag.Tag.readNamedTag(Tag.java:91) ~[server.jar:?]
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:93) ~[server.jar:?]
        at cn.nukkit.nbt.NBTIO.read(NBTIO.java:89) ~[server.jar:?]
        at cn.nukkit.nbt.NBTIO.readCompressed(NBTIO.java:127) ~[server.jar:?]
        at cn.nukkit.level.format.generic.BaseLevelProvider.<init>(BaseLevelProvider.java:58) ~[server.jar:?]
        at cn.nukkit.level.format.anvil.Anvil.<init>(Anvil.java:41) ~[server.jar:?]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_272]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_272]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_272]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_272]
        at cn.nukkit.level.Level.lambda$new$1(Level.java:312) ~[server.jar:?]
        ... 6 more
18:39:37 [WARN ] No level called "nether" found or nether is disabled in server properties! Nether functionality will be disabled.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ Click the link above to see the future.
 ### Fixes
 - _Alpha Bug_: Fixes entities from the MobPlugin going into the ground (compatibility issue)
 - [#770] Bamboo not dropping when broken, were also affecting blocks with id > 255
+- [#765] Unsafe level.dat writes could lead to world corruption
 - [#766] Error saving region files with content over 2 GB
 - [#776] _Alpha Bug_: Grindstone can't remove enchants
 - [#777] Falling block falling though scaffolds
@@ -21,6 +22,7 @@ Click the link above to see the future.
 - [#778] Unable to get `minecraft:mob_spawner` with `/give Nick mob_spawner`
 
 ### Changed
+- [#765] The `ServerBrand` tag in the `level.dat` file will be set to `PowerNukkit` now 
 - [#776] Grindstone won't reset the repair cost anymore
 
 ## 1.3.2.0-PN-ALPHA.2 - 2020-11-21
@@ -630,6 +632,7 @@ Fixes several anvil issues.
 [#469]: https://github.com/PowerNukkit/PowerNukkit/issues/469
 [#475]: https://github.com/PowerNukkit/PowerNukkit/issues/475
 [#544]: https://github.com/PowerNukkit/PowerNukkit/issues/544
+[#765]: https://github.com/PowerNukkit/PowerNukkit/issues/765
 [#766]: https://github.com/PowerNukkit/PowerNukkit/issues/766
 [#770]: https://github.com/PowerNukkit/PowerNukkit/issues/770
 [#776]: https://github.com/PowerNukkit/PowerNukkit/issues/776

--- a/src/main/java/cn/nukkit/level/format/anvil/Anvil.java
+++ b/src/main/java/cn/nukkit/level/format/anvil/Anvil.java
@@ -14,11 +14,11 @@ import cn.nukkit.scheduler.AsyncTask;
 import cn.nukkit.utils.BinaryStream;
 import cn.nukkit.utils.ChunkException;
 import cn.nukkit.utils.ThreadCache;
+import cn.nukkit.utils.Utils;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
+import lombok.extern.log4j.Log4j2;
 
-import java.io.File;
-import java.io.FileOutputStream;
-import java.io.IOException;
+import java.io.*;
 import java.nio.ByteOrder;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -30,6 +30,7 @@ import java.util.regex.Pattern;
  * author: MagicDroidX
  * Nukkit Project
  */
+@Log4j2
 public class Anvil extends BaseLevelProvider {
     public static final int VERSION = 19133;
     static private final byte[] PAD_256 = new byte[256];
@@ -68,8 +69,9 @@ public class Anvil extends BaseLevelProvider {
     }
 
     public static void generate(String path, String name, long seed, Class<? extends Generator> generator, Map<String, String> options) throws IOException {
-        if (!new File(path + "/region").exists()) {
-            new File(path + "/region").mkdirs();
+        File regionDir = new File(path + "/region");
+        if (!regionDir.exists() && !regionDir.mkdirs()) {
+            throw new IOException("Could not create the directory "+regionDir);
         }
 
         CompoundTag levelData = new CompoundTag("Data")
@@ -96,7 +98,13 @@ public class Anvil extends BaseLevelProvider {
                 .putLong("Time", 0)
                 .putLong("SizeOnDisk", 0);
 
-        NBTIO.writeGZIPCompressed(new CompoundTag().putCompound("Data", levelData), new FileOutputStream(path + "level.dat"), ByteOrder.BIG_ENDIAN);
+        Utils.safeWrite(new File(path, "level.dat"), file -> {
+            try(FileOutputStream fos = new FileOutputStream(file); BufferedOutputStream out = new BufferedOutputStream(fos)) {
+                NBTIO.writeGZIPCompressed(new CompoundTag().putCompound("Data", levelData), out, ByteOrder.BIG_ENDIAN);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
     }
 
     @Override

--- a/src/main/java/cn/nukkit/level/format/generic/BaseLevelProvider.java
+++ b/src/main/java/cn/nukkit/level/format/generic/BaseLevelProvider.java
@@ -1,5 +1,6 @@
 package cn.nukkit.level.format.generic;
 
+import cn.nukkit.Nukkit;
 import cn.nukkit.Server;
 import cn.nukkit.level.GameRules;
 import cn.nukkit.level.Level;
@@ -9,6 +10,8 @@ import cn.nukkit.level.generator.Generator;
 import cn.nukkit.math.Vector3;
 import cn.nukkit.nbt.NBTIO;
 import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.nbt.tag.ListTag;
+import cn.nukkit.nbt.tag.StringTag;
 import cn.nukkit.utils.ChunkException;
 import cn.nukkit.utils.LevelException;
 import cn.nukkit.utils.Utils;
@@ -92,6 +95,8 @@ public abstract class BaseLevelProvider implements LevelProvider {
         if (!this.levelData.contains("generatorOptions")) {
             this.levelData.putString("generatorOptions", "");
         }
+        
+        this.levelData.putList(new ListTag<>("ServerBrand").add(new StringTag("", Nukkit.CODENAME)));
 
         this.spawn = new Vector3(this.levelData.getInt("SpawnX"), this.levelData.getInt("SpawnY"), this.levelData.getInt("SpawnZ"));
     }

--- a/src/main/java/cn/nukkit/level/format/generic/BaseLevelProvider.java
+++ b/src/main/java/cn/nukkit/level/format/generic/BaseLevelProvider.java
@@ -11,15 +11,14 @@ import cn.nukkit.nbt.NBTIO;
 import cn.nukkit.nbt.tag.CompoundTag;
 import cn.nukkit.utils.ChunkException;
 import cn.nukkit.utils.LevelException;
+import cn.nukkit.utils.Utils;
 import com.google.common.collect.ImmutableMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
+import lombok.extern.log4j.Log4j2;
 
-import java.io.File;
-import java.io.FileInputStream;
-import java.io.FileOutputStream;
-import java.io.IOException;
+import java.io.*;
 import java.nio.ByteOrder;
 import java.util.HashMap;
 import java.util.Map;
@@ -29,6 +28,7 @@ import java.util.concurrent.atomic.AtomicReference;
  * author: MagicDroidX
  * Nukkit Project
  */
+@Log4j2
 public abstract class BaseLevelProvider implements LevelProvider {
     protected Level level;
 
@@ -49,11 +49,36 @@ public abstract class BaseLevelProvider implements LevelProvider {
     public BaseLevelProvider(Level level, String path) throws IOException {
         this.level = level;
         this.path = path;
-        File file_path = new File(this.path);
-        if (!file_path.exists()) {
-            file_path.mkdirs();
+        File filePath = new File(this.path);
+        if (!filePath.exists() && !filePath.mkdirs()) {
+            throw new LevelException("Could not create the directory "+filePath);
         }
-        CompoundTag levelData = NBTIO.readCompressed(new FileInputStream(new File(this.getPath() + "level.dat")), ByteOrder.BIG_ENDIAN);
+
+        CompoundTag levelData;
+        File levelDatFile = new File(getPath(), "level.dat");
+        try (FileInputStream fos = new FileInputStream(levelDatFile); BufferedInputStream input = new BufferedInputStream(fos)) {
+            levelData = NBTIO.readCompressed(input, ByteOrder.BIG_ENDIAN);;
+        } catch (Exception e) {
+            log.fatal("Failed to load the level.dat file at "+levelDatFile.getAbsolutePath()+", attempting to load level.dat_old instead!");
+            try {
+                File old = new File(getPath(), "level.dat_old");
+                if (!old.isFile()) {
+                    log.fatal("The file "+old.getAbsolutePath()+" does not exists!");
+                    throw new FileNotFoundException("The file "+old.getAbsolutePath()+" does not exists!");
+                }
+                try (FileInputStream fos = new FileInputStream(old); BufferedInputStream input = new BufferedInputStream(fos)) {
+                    levelData = NBTIO.readCompressed(input, ByteOrder.BIG_ENDIAN);
+                } catch (Exception e2) {
+                    log.fatal("Failed to load the level.dat_old file at "+levelDatFile.getAbsolutePath());
+                    throw e2;
+                }
+            } catch (Exception e2) {
+                LevelException ex = new LevelException("Could not load the level.dat and the level.dat_old files. You might need to restore them from a backup!", e);
+                ex.addSuppressed(e2);
+                throw ex;
+            }
+        }
+        
         if (levelData.get("Data") instanceof CompoundTag) {
             this.levelData = levelData.getCompound("Data");
         } else {
@@ -303,10 +328,18 @@ public abstract class BaseLevelProvider implements LevelProvider {
 
     @Override
     public void saveLevelData() {
+        File levelDataFile = new File(getPath(), "level.dat");
         try {
-            NBTIO.writeGZIPCompressed(new CompoundTag().putCompound("Data", this.levelData), new FileOutputStream(this.getPath() + "level.dat"));
+            Utils.safeWrite(levelDataFile, file -> {
+                try(FileOutputStream fos = new FileOutputStream(file); BufferedOutputStream out = new BufferedOutputStream(fos)) {
+                    NBTIO.writeGZIPCompressed(new CompoundTag().putCompound("Data", this.levelData), out);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            });
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            log.fatal("Failed to save the level.dat file at "+levelDataFile.getAbsolutePath());
+            throw new UncheckedIOException(e);
         }
     }
 

--- a/src/main/java/cn/nukkit/utils/LevelException.java
+++ b/src/main/java/cn/nukkit/utils/LevelException.java
@@ -1,5 +1,8 @@
 package cn.nukkit.utils;
 
+import cn.nukkit.api.PowerNukkitOnly;
+import cn.nukkit.api.Since;
+
 /**
  * author: MagicDroidX
  * Nukkit Project
@@ -7,5 +10,11 @@ package cn.nukkit.utils;
 public class LevelException extends ServerException {
     public LevelException(String message) {
         super(message);
+    }
+
+    @PowerNukkitOnly
+    @Since("1.3.2.0-PN")
+    public LevelException(String message, Throwable cause) {
+        super(message, cause);
     }
 }

--- a/src/main/java/cn/nukkit/utils/Utils.java
+++ b/src/main/java/cn/nukkit/utils/Utils.java
@@ -2,6 +2,7 @@ package cn.nukkit.utils;
 
 import cn.nukkit.api.PowerNukkitOnly;
 import cn.nukkit.api.Since;
+import lombok.extern.log4j.Log4j2;
 
 import java.io.*;
 import java.lang.management.ManagementFactory;
@@ -13,10 +14,12 @@ import java.util.Map;
 import java.util.SplittableRandom;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
 
 /**
  * @author MagicDroidX (Nukkit Project)
  */
+@Log4j2
 public class Utils {
     @PowerNukkitOnly
     @Since("1.3.2.0-PN")
@@ -25,6 +28,45 @@ public class Utils {
     @PowerNukkitOnly
     @Since("1.3.2.0-PN")
     public static final SplittableRandom random = new SplittableRandom();
+    
+    @PowerNukkitOnly
+    @Since("1.3.2.0-PN")
+    public static void safeWrite(File currentFile, Consumer<File> operation) throws IOException {
+        File parent = currentFile.getParentFile();
+        File newFile = new File(parent, currentFile.getName()+"_new");
+        File oldFile = new File(parent, currentFile.getName()+"_old");
+        File olderFile = new File(parent, currentFile.getName()+"_older");
+
+        if (olderFile.isFile() && !olderFile.delete()) {
+            log.fatal("Could not delete the file "+olderFile.getAbsolutePath());
+        }
+
+        if (newFile.isFile() && !newFile.delete()) {
+            log.fatal("Could not delete the file "+newFile.getAbsolutePath());
+        }
+        
+        try {
+            operation.accept(newFile);
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+        
+        if (oldFile.isFile()) {
+            if (olderFile.isFile()) {
+                Utils.copyFile(oldFile, olderFile);
+            } else if (!oldFile.renameTo(olderFile)) {
+                throw new IOException("Could not rename the " + oldFile + " to " + olderFile);
+            }
+        }
+
+        if (currentFile.isFile() && !currentFile.renameTo(oldFile)) {
+            throw new IOException("Could not rename the " + currentFile + " to " + oldFile);
+        }
+
+        if (!newFile.renameTo(currentFile)) {
+            throw new IOException("Could not rename the " + newFile + " to " + currentFile);
+        }
+    }
 
     public static void writeFile(String fileName, String content) throws IOException {
         writeFile(fileName, new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8)));

